woocommerce/woocommerce
# [Settings] Add transformer for grouping settings elements into sections and checkbox groups

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes https://github.com/woocommerce/woocommerce/issues/52798

This PR introduces a transformer class that organizes settings data into logical groups while maintaining the original structure and order of settings by:

- Grouping settings between `title` and `sectionend` into logical groups
- Grouping consecutive checkboxes marked with `checkboxgroup` attributes
- Preserving original settings when grouping conditions aren't met
- Maintaining the order of settings within sections

For simplicity, this transformer does not handle nested groups since we don't have any use cases for it yet.

I implemented it in PHP because we can cache the transformed data on server side to improve performance instead of transforming the data on-the-fly in JS, which delays the initial page load. But this PR doesn't include that to keep this PR focused on the main task.

Changes:
- Added `Transformer` class for settings data transformation. 
- Update `global.d.ts`
- Add unit tests

Input example:

```php
[
    'tab_id' => [
        'sections' => [
            'section_id' => [
                'settings' => [
                    ['type' => 'title', 'id' => 'group_1', ...],
                    ['type' => 'text', ...],
                    ['type' => 'sectionend', 'id' => 'group_1']
                ]
            ]
        ]
    ]
]
```

Output example:

```php
[
    'tab_id' => [
        'sections' => [
            'section_id' => [
                'settings' => [
                    [
                        'type' => 'group',
                        'id' => 'group_1',
                        'settings' => [
                            ['type' => 'text', ...]
                        ]
                    ]
                ]
            ]
        ]
    ]
]
```



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Review the code changes and confirm unit tests cover all possible cases.
2. Create a fresh site with `settings` feature flag enabled.
3. Go to the settings page and open dev tools.
4. Run `window.wcSettings.admin.settingsData` in the console and confirm the structure is as expected.


See that Products settings for example is grouped as expected.

![Screenshot 2024-11-19 at 10 28 23](https://github.com/user-attachments/assets/6e6e9f4d-fac7-4d6b-8b0a-d05ca731b891)
![Screenshot 2024-11-19 at 10 30 01](https://github.com/user-attachments/assets/d94423ed-3d8d-423c-aab4-1a7465cef124)




### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
